### PR TITLE
fix: "tracestate" parsing must allow whitespace around list-members

### PR DESCRIPTION
**Pull request by trentm**
Per https://w3c.github.io/trace-context/#tracestate-header-field-values
the tracestate header can have whitespace around list-members. E.g.
    'vendor1=foo, vendor2=bar ,\t vendor3=baz '

Refs: https://github.com/elastic/apm-agent-rum-js/issues/1334

---

I'm fairly confident on the regex changes, but not at all on possible wider impact. I haven't followed closely how `_parse_tracestate` is used in the agent code.

### CODE REVIEW

File: elasticapm/utils/disttracing.py [hunk 1]
```diff
@@ -208,7 +212,7 @@ def _parse_tracestate(self, tracestate) -> Dict[str, str]:
 
         ret = {}
         try:
-            state = re.search(r"(?:,|^)es=([^,]*)", tracestate).group(1).split(";")
+            state = re.search(r"(?:,|^)\s*es=([^,]*?)\s*(?:,|$)", tracestate).group(1).split(";")
```
Comment (basepi) [inline elasticapm/utils/disttracing.py:215]: Is this question mark superfluous? The `*` already makes it optional, yes?

```suggestion
            state = re.search(r"(?:,|^)\s*es=([^,]*)\s*(?:,|$)", tracestate).group(1).split(";")
```

**Review by basepi — APPROVED** [review]

**Comment by basepi** [comment]
@trentm it won't let me ask you for review since you're the author...what do you think of my changes?

File: elasticapm/utils/disttracing.py (hunk 1 above)
Comment (trentm) [reply elasticapm/utils/disttracing.py]: That `?` is for a non-greedy match, otherwise, I think (I'd need to test), the `([^,]*)` could slurp up the trailing whitespace rather than leaving it to the `\s*`.

Comment (trentm) [reply elasticapm/utils/disttracing.py]:
```
>>> tracestate = 'foo=bar, es=blah , bing=bang '
>>> re.search(r"(?:,|^)\s*es=([^,]*)\s*(?:,|$)", tracestate).group(1)
'blah '
>>> re.search(r"(?:,|^)\s*es=([^,]*?)\s*(?:,|$)", tracestate).group(1)
'blah'
```

Comment (basepi) [reply elasticapm/utils/disttracing.py]: Ah, right, thank you. In that case I can undo my "fix" in https://github.com/elastic/apm-agent-python/pull/1728/commits/a609b1a52df7dded2eb7a27720ef223daa875f54

Comment (trentm) [reply elasticapm/utils/disttracing.py:215]: Ah, I hadn't see the `.strip()` call. I think the issue with that if this were for **general** tracestate parsing, is that a `value` here can have *leading* whitespace (per https://www.w3.org/TR/trace-context/#value). So it would have to be a `.rstrip()` if kept.   I'm nitpicking though, because IIUC the Python impl *only* ever processing the `es` vendor value and we'll never care about leading whitespace.

Comment (trentm) [reply elasticapm/utils/disttracing.py:215]: A possible alternative for parsing this all, if the regex is deemed getting too complex, would be to:
- `str.find('es=')`
- if there was a hit, then `.split(',')` and `.trim()` each entry
- then pick the one(s) starting with 'es='.

**Comment by trentm** [comment]
@basepi Should I merge this? Or should  there be a changelog entry or something?
I'm away tomorrow, so feel free to take this one over.

File: elasticapm/utils/disttracing.py (hunk 1 above)
Comment (basepi) [reply elasticapm/utils/disttracing.py:215]: I think it's fine. But thanks for the thoughtful suggestions.
